File: modules/zon/src/main/java/com/falsepattern/zigbrains/zon/folding/ZonFoldingBuilder.java
/*
 * Copyright 2023-2024 FalsePattern
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.falsepattern.zigbrains.zon.folding;

import com.falsepattern.zigbrains.zon.psi.ZonStruct;
import com.falsepattern.zigbrains.zon.psi.ZonVisitor;
import com.intellij.lang.ASTNode;
import com.intellij.lang.folding.CustomFoldingBuilder;
import com.intellij.lang.folding.FoldingDescriptor;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.project.DumbAware;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class ZonFoldingBuilder extends CustomFoldingBuilder implements DumbAware {
    @Override
    protected void buildLanguageFoldRegions(@NotNull List<FoldingDescriptor> descriptors, @NotNull PsiElement root, @NotNull Document document, boolean quick) {
        root.accept(new ZonVisitor() {
            @Override
            public void visitElement(@NotNull PsiElement element) {
                super.visitElement(element);
                element.acceptChildren(this);
            }

            @Override
            public void visitStruct(@NotNull ZonStruct o) {
                super.visitStruct(o);
                descriptors.add(new FoldingDescriptor(o, o.getTextRange()));
            }
        });
    }

    @Override
    protected String getLanguagePlaceholderText(@NotNull ASTNode node, @NotNull TextRange range) {
        return ".{...}";
    }

    @Override
    protected boolean isRegionCollapsedByDefault(@NotNull ASTNode node) {
        return false;
    }
}

File: modules/zon/src/main/kotlin/com/falsepattern/zigbrains/zon/folding/ZonFoldingBuilder.kt
package com.falsepattern.zigbrains.zon.folding

import com.falsepattern.zigbrains.zon.psi.ZonStruct
import com.falsepattern.zigbrains.zon.psi.ZonVisitor
import com.intellij.lang.ASTNode
import com.intellij.lang.folding.CustomFoldingBuilder
import com.intellij.lang.folding.FoldingDescriptor
import com.intellij.openapi.editor.Document
import com.intellij.openapi.project.DumbAware
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement

class ZonFoldingBuilder: CustomFoldingBuilder(), DumbAware {
    override fun buildLanguageFoldRegions(
        descriptors: MutableList<FoldingDescriptor>,
        root: PsiElement,
        document: Document,
        quick: Boolean
    ) {
        root.accept(object: ZonVisitor() {
            override fun visitElement(element: PsiElement) {
                super.visitElement(element)
                element.acceptChildren(this)
            }

            override fun visitStruct(o: ZonStruct) {
                super.visitStruct(o)
                descriptors.add(FoldingDescriptor(o, o.textRange))
            }
        })
    }

    override fun getLanguagePlaceholderText(node: ASTNode, range: TextRange) = ".{...}"

    override fun isRegionCollapsedByDefault(node: ASTNode) = false
}